### ISSUE TITLE
Question about the model's input and output

### ISSUE THREAD

**Issue by sonnefred** [NVlabs/alpamayo#14]
Hi, thanks for sharing the work. I have some questions about the model's input and output. 
1. When loading the dataset, why set the t0 to 5100000? What's the exact meaning of this parameter? If I change it to 0, what will it affect?

![Image](https://github.com/user-attachments/assets/ad8cf8e0-c907-4458-b0e0-84e4d6991ac1)

2. When I input 10 frames and 50 frames into the model, what's the difference between its output (the 64 predicted waypoints)?
Looking forward to your reply, thanks!

**Comment by super-anova** [comment]
Hi, thanks for your question.

t0_us defines the timestamp in microseconds at which to load the data from the physical AI AV dataset. For example, say if t0_us=5.1s seconds, num_history_steps=16, num_future_steps=64, this function would load the (3.5s, 11.5s] of the trajectory data. See [these comments](https://github.com/NVlabs/alpamayo/blob/main/src/alpamayo_r1/load_physical_aiavdataset.py#L103-L104).

You can't set t0_us to 0 since there is no history to load at time zero.

Also note that the number of camera frames to load is controlled by another parameter: `num_frames`

**Comment by lonexreb** [comment]
Mapping out the IO concretely from the released code. (References are to files at `main`.)

### Inputs to `AlpamayoR1.sample_trajectories_from_data_with_vlm_rollout`

The model consumes two things — a **tokenized text/image prompt** and **ego-motion history tensors**.

**1. Camera/text prompt (built by `helper.create_message`, `src/alpamayo_r1/helper.py:28-68`):**
- 4 camera views × 4 frames each (default) → `image_frames` tensor of shape `(N_cameras, num_frames, 3, H, W)`, produced by `load_physical_aiavdataset` (`src/alpamayo_r1/load_physical_aiavdataset.py:208-219`). Default cameras: `CAMERA_CROSS_LEFT_120FOV`, `CAMERA_FRONT_WIDE_120FOV`, `CAMERA_CROSS_RIGHT_120FOV`, `CAMERA_FRONT_TELE_30FOV`.
- A chat-template prompt with system/user/assistant roles, embedded image tokens, a 48-token history-trajectory placeholder (`<|traj_history_start|>` + 48 × `<|traj_history|>` + `<|traj_history_end|>`), and an instruction to output chain-of-causation reasoning before the future trajectory.
- The placeholder pad tokens get replaced at runtime by trajectory tokens in `fuse_traj_tokens` / `tokenize_history_trajectory` (`src/alpamayo_r1/models/base_model.py:91-122`), which encodes `ego_history_xyz` + `ego_history_rot` through `traj_tokenizer` and substitutes them into `input_ids`.

**2. Ego history tensors:**
- `ego_history_xyz`: `(B, 1, num_history_steps, 3)` — default `num_history_steps=16`, i.e. 1.6 s at 10 Hz.
- `ego_history_rot`: `(B, 1, num_history_steps, 3, 3)` — rotation matrices in the local frame at `t0`.

Both are transformed to the ego frame at `t0` before tokenization (`load_physical_aiavdataset.py:128-144`).

### Outputs

`sample_trajectories_from_data_with_vlm_rollout` returns `(pred_xyz, pred_rot, extra)`:

- `pred_xyz`: `(B, num_traj_sets, num_traj_samples, num_future_steps, 3)` — XYZ in the local frame at `t0`. Default `num_future_steps=64` (6.4 s at 10 Hz).
- `pred_rot`: `(B, num_traj_sets, num_traj_samples, num_future_steps, 3, 3)` — rotation matrices.
- `extra` (when `return_extra=True`): dict with text-token decodings, including `extra["cot"]` → the chain-of-causation reasoning string per sample. See `extract_text_tokens` and the rearrange to `[B, ns, nj]` at `alpamayo_r1.py:322-329`.

### Two-stage flow

1. **VLM rollout** (`alpamayo_r1.py:194-208`): `self.vlm.generate(...)` runs autoregressive generation with `ExpertLogitsProcessor` (masks all 768 trajectory token IDs to `-inf` so the VLM cannot accidentally emit them) and `StopAfterEOS(eos_token_id=<traj_future_start>)`. The VLM emits CoC reasoning, optional meta-action text, and stops one token after `<traj_future_start>`.
2. **Diffusion expert** (`alpamayo_r1.py:256-299`): the VLM's KV cache is reused as a prefix; `step_fn` projects noisy actions through `action_in_proj`, runs the expert decoder on the cached prefill, projects back through `action_out_proj`, and `self.diffusion.sample(...)` does flow-matching denoising. The final action vector is converted to `(pred_xyz, pred_rot)` via `self.action_space.action_to_traj` (`alpamayo_r1.py:309-311`).

### Quick reference

| Field | Shape | File:Line |
|---|---|---|
| `image_frames` | `(N_cam, T, 3, H, W)` | `load_physical_aiavdataset.py:209` |
| `ego_history_xyz` | `(B, 1, 16, 3)` | `:211` |
| `ego_history_rot` | `(B, 1, 16, 3, 3)` | `:212` |
| `pred_xyz` | `(B, ns, nj, 64, 3)` | `alpamayo_r1.py:314-316` |
| `pred_rot` | `(B, ns, nj, 64, 3, 3)` | `alpamayo_r1.py:317-319` |
| `extra["cot"]` | `np.ndarray[B, ns, nj]` of strings | `alpamayo_r1.py:322-328` |

`src/alpamayo_r1/test_inference.py` is the smallest end-to-end example showing all of the above.

**Comment by lucamodica** [comment]
@lonexreb thanks a lot for breaking down I/O of the model : ).

@super-anova I was wondering about the number of frames in training though. Both in the inference and fine-tuning code I have notice that 4 frames are always loaded, and in the paper It is not explicitly specified. Are also the camera at 10Hz as for the history trajectory?

**Comment by super-anova** [comment]
@lucamodica It is 4 cameras and 4 frames per each camera at 10Hz. So that is 0.4 seconds of camera videos. That is the model input during training.

The 16 points from ego trajectory is also at 10Hz.

**Comment by lucamodica** [comment]
@super-anova thanks you! Considering the 1.6 seconds history, are these 4 frames uniformly sampled throughout the history time or is it just the last 0.4s history as in the inference demo script?

**Comment by super-anova** [comment]
@lucamodica It is the last 0.4s history as demo'd in the inference script (not uniformly sampled from the 1.6s ego trajectory history).

**Comment by lonexreb** [comment]
@lucamodica For training the 4 frames cover only the **last 0.4s** of the 1.6s history window — same as the inference demo, not uniformly subsampled across history.

The sampling is hard-coded in `load_physical_aiavdataset` (`src/alpamayo_r1/load_physical_aiavdataset.py:164-168`):

```python
# Image timestamps: if num_frames=4, load at [t0-0.3s, t0-0.2s, t0-0.1s, t0]
image_timestamps = np.array(
    [t0_us - (num_frames - 1 - i) * int(time_step * 1_000_000) for i in range(num_frames)],
    dtype=np.int64,
)
```

With defaults `num_frames=4`, `time_step=0.1s` → `[t0-0.3, t0-0.2, t0-0.1, t0]`. The camera stride equals `time_step` (i.e. 10 Hz), so cameras and the 16-step ego history share the same rate but cover different durations (0.4s vs 1.6s).

The SFT/RL dataloaders use this same function — see `src/alpamayo_r1/data/pai.py:106-113`, which calls `load_physical_aiavdataset(...)` with the same `time_step`/`num_history_steps`. There is no separate training-time frame-sampling path, so the earlier 1.2s of history is represented only by the 16 ego-pose tokens, not by camera frames.

**Comment by wyddmw** [comment]
Hi all, I would like to ask about the high-level navigation command. According to the current released model, the planned trajectory and cot reasoning are not conditioned on any high-level navigation command, which is not consistent with the demonstration in the paper. I would like to ask how to get access to the mentioned high-level commands as input. Are they included in the current release coc labels?
